spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/LazyScopePassingSpanSubscriber.java

```diff
@@ -29,48 +29,67 @@
  * @author Marcin Grzejszczak
  * @since 2.0.0
  */
-// TODO: why are we extending AtomicBoolean and not actually using its methods?
-final class LazySpanSubscriber<T> extends AtomicBoolean implements SpanSubscription<T> {
+final class LazyScopePassingSpanSubscriber<T> implements SpanSubscription<T> {
+
+	final CoreSubscriber<? super T> subscriber;
+
+	final Context context;
+
+	final CurrentTraceContext currentTraceContext;
 
-	private final Supplier<SpanSubscription<T>> supplier;
+	SpanSubscription<T> spanSubscription;
 
-	LazySpanSubscriber(Supplier<SpanSubscription<T>> supplier) {
-		this.supplier = supplier;
+	LazyScopePassingSpanSubscriber(CoreSubscriber<? super T> subscriber, Context context,
+			CurrentTraceContext currentTraceContext) {
+		this.subscriber = subscriber;
+		this.context = context;
+		this.currentTraceContext = currentTraceContext;
 	}
 
 	@Override
 	public void onSubscribe(Subscription subscription) {
-		this.supplier.get().onSubscribe(subscription);
+		spanSubscription().onSubscribe(subscription);
 	}
 
 	@Override
 	public void request(long n) {
-		this.supplier.get().request(n);
+		spanSubscription().request(n);
 	}
 
 	@Override
 	public void cancel() {
-		this.supplier.get().cancel();
+		spanSubscription().cancel();
 	}
 
 	@Override
 	public void onNext(T o) {
-		this.supplier.get().onNext(o);
+		spanSubscription().onNext(o);
 	}
 
 	@Override
 	public void onError(Throwable throwable) {
-		this.supplier.get().onError(throwable);
+		spanSubscription().onError(throwable);
 	}
 
 	@Override
 	public void onComplete() {
-		this.supplier.get().onComplete();
+		spanSubscription().onComplete();
 	}
 
 	@Override
 	public Context currentContext() {
-		return this.supplier.get().currentContext();
+		return spanSubscription().currentContext();
+	}
+
+	SpanSubscription<T> spanSubscription() {
+		if (this.spanSubscription == null) {
+			TraceContext parent = ReactorSleuth.traceContext(this.context,
+					this.currentTraceContext);
+
+			this.spanSubscription = new ScopePassingSpanSubscriber<>(this.subscriber,
```

Comment:
TODO: we can doublecheck lazycurrentTraceContext.unwrap here to straight pipe future operations

Reply:
Maybe worth a comment that this object is safe to allocate multiple times so we don't worry about thread safety (at least assuming that's correct ;)

Reply:
I don't know if it is, the class was formerly allocated each time :D maybe someone else can chime in on this one.

Reply:
there's no obvious state concern, I guess I mean I'm unaware of a subtle one.

Reply:
actually I think it is less distracting to assume this should be provisioned only once.. like a normal lazy. I tried hunting down the various interfaces and there could be different threads invoking callbacks as far as I can tell.